lib/doubly_linked_list.rb

```diff
@@ -0,0 +1,113 @@
+class DoublyNode
+    attr_reader :data
+    attr_accessor :next, :previous
+    
+    def initialize(value)
+        @data = value
+        @next = nil
+        @previous = nil
+    end
+end
+
+class DoublyLinkedList
+    def initialize
+        @head = nil
+        @tail = nil
+        @size = 0
+    end
+    
+    def length
+        return @size
+        # raise NotImplementedError
+    end
+    
+    def add_first(value)
+        new_node = DoublyNode.new(value)
+        
+        if @head
+            @head.previous = new_node
+            new_node.next = @head
+        else
+            @tail = new_node
+        end
+        @head = new_node
+        @size += 1
+    end
+    
+    def add_last(value)
+        new_node = DoublyNode.new(value)
+        
+        if @tail
+            @tail.next = new_node
+            new_node.previous = @tail
+        else
+            @head = new_node
+        end
+        @tail = new_node
+        @size += 1
+    end
+    
+    def get_first
+        return @head.data if @head
+        return @head
+    end
+    
+    def get_last
+        return @tail.data if @tail
+        return @tail
```

Comment:
This method works, but if `@tail` is `nil`, I suggest explicitly returning `nil` here.

```suggestion
        return nil
```